### PR TITLE
Handles the empty string encoding/decoding in Python 2.7

### LINKED ISSUE #366: ExtendedTests failing on Mac Python 2.7 build due to NGramFeaturizer.py
**Describe the bug**
The Mac Py 2.7 build is behaving differently then the Linux Py 2.7 builds in that the results obtained in test_ngramfeaturizer.py on Linux is different from the results obtained in the respective Mac counterpart. Running it on the Linux Py 2.7 builds provide the correct outputs, but not on the Mac Py 2.7 build. This issue became apparent after the fixing of #360.

Link to exact failure point of bug:
https://dev.azure.com/aifx/public/_build/results?buildId=1979&view=logs&j=647a623b-a2e8-58c1-2f61-78ff08c24c44&t=bc9337da-f88d-5297-6b65-82f9ec31f5a3&l=5059

**Desktop (please complete the following information):**
 - OS: MacOS
 - Python Version: 2.7

### PR DESCRIPTION
**Pull request by mstfbl**
Additional fix to Issue #360 
Fixes #366 

Handles the case where, with Python 2.7, string s is valid with utf_8 encoding, but is the empty string. I discovered the need for this fix as 2 test cases that were skipped before PR #359 were failing even with PR #359, due to this case with the empty string on Python 2.7.

These tests were:
- TestNGramFeaturizer.test_ngramfeaturizer()
- TestLightGbmClassifier.test_lightgbmclassifier()

**EDIT**:  My observation on the error provided in TestLightGbmClassifier.test_lightgbmclassifier() was incorrect. The error is only visible in TestNGramFeaturizer.test_ngramfeaturizer(). The exact failure is [here](https://dev.azure.com/aifx/public/_build/results?buildId=1976&view=logs&j=d602e5ee-b117-53e9-baca-06d38555e9ae&t=7a4bfe99-128e-548b-1b46-3dfd9c68bb4f&l=4393) and [here](https://dev.azure.com/aifx/public/_build/results?buildId=1976&view=logs&j=d602e5ee-b117-53e9-baca-06d38555e9ae&t=7a4bfe99-128e-548b-1b46-3dfd9c68bb4f&l=5057), shown on the Mac 2.7 build.

### CODE REVIEW

**Comment by ganik** [comment]
>         if (bp::extract<const char*>(str(s).encode("utf_8")).check())

What does this check() method do?

---
Refers to: src/NativeBridge/DataViewInterop.h:241 in 1f92b8a. [](commit_id = 1f92b8a790dfab987ca763ec98c176102aabc368, deletion_comment = False)

File: src/NativeBridge/DataViewInterop.h [hunk 1]
```diff
@@ -243,15 +243,23 @@ class DataSourceBlock
             size = -1;
             missing = -1;
             pch = bp::extract<const char*>(str(s).encode("utf_8"));
+            if (str(s).encode("utf_8").is_none())
+            {
+                size = 0;
```
Comment (ganik) [inline src/NativeBridge/DataViewInterop.h]: >size = 0; [](start = 16, length = 9)

When do we come here?

**Comment by mstfbl** [comment]
The check() method in this context is:

```
  template <class Ptr>
  inline bool extract_pointer<Ptr>::check() const
  {
      return m_source == Py_None || m_result != 0;
  }
```
According to the Boost docs [here](https://www.boost.org/doc/libs/1_66_0/libs/python/doc/html/reference/to_from_python_type_conversion.html), I believe this functions checks whether the given object can be converted to a given C# type, which in this case would be <const char*>. This check is necessary to properly handle a given text from Python to C++, and later to C#.

File: src/NativeBridge/DataViewInterop.h (hunk 1 above)
Comment (mstfbl) [reply src/NativeBridge/DataViewInterop.h]: For context, the `is_none `function and its concerning values:

```
inline bool api::object_base::is_none() const
{
    return (m_ptr == Py_None);
}
```
`PyObject* m_ptr;
`
```
/*
_Py_NoneStruct is an object of undefined type which can be used in contexts
where NULL (nil) is not suitable (since NULL often means 'error').

Don't forget to apply Py_INCREF() when returning this value!!!
*/
PyAPI_DATA(PyObject) _Py_NoneStruct; /* Don't use this directly */
#define Py_None (&_Py_NoneStruct)
```
I believe the code reaches in this if statement in line 246 if the given object is equal to the `NullType `in Python.
The peculiar question is, on Linux builds this statement works fine, but in our Mac builds, this statement doesn't fix the failure. Perhaps `Py_None` doesn't exactly point to  `NoneType `on Mac? In addition, this test was being skipped until PR #359, so this means this issue did not surface until this test was no longer being skipped.

**Comment by mstfbl** [comment]
As of commit [d9b9544](https://github.com/microsoft/NimbusML/pull/362/commits/d9b9544f692402c2a490d9f25aab741a9feafe01), the Mac Py2.7 build passes.

**Comment by mstfbl** [comment]
Closing this PR for now as we are temporarily disabling the corresponding tests for the Mac Python 2.7 builds with PR #391 .
